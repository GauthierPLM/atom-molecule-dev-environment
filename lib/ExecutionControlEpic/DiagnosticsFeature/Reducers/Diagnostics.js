"use babel";
// @flow

import type { Diagnostic } from "../Types/types.js.flow";
import { fromJS, List } from "immutable";

export default function Diagnostics(
  state: DiagnosticsReducer = List(),
  action: any
): DiagnosticsReducer {
  switch (action.type) {
    case "ADD_DIAGNOSTIC_FOR_TASK":
      return state.push(action.payload.diagnostic);
    case "ADD_DIAGNOSTICS_FOR_TASK":
      return state.concat(action.payload.diagnostics);
    case "REMOVE_DIAGNOSTICS_OF_TASK":
      return state.filter(diagnostic => diagnostic.task != action.payload.task);
    default:
      return state;
  }
}

export type DiagnosticsReducer = List<Diagnostic>;
