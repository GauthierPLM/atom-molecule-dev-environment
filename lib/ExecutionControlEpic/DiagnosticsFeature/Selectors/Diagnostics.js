"use babel";
// @flow
import type { DiagnosticsReducer } from "../Reducers/Diagnostics";
import type { Diagnostic } from "../Types/types.js.flow";
import { List } from "immutable";

export function selectDiagnosticsOfTask(
  diagnostics: DiagnosticsReducer,
  taskId: string,
): List<Diagnostic> {
  return diagnostics.filter(diagnostic => diagnostic.task == taskId);
}

export function selectDiagnosticsOfStep(
  diagnostics: List<Diagnostic>,
  step: number,
) {
  return diagnostics.filter(diagnostic => diagnostic.step == step);
}

export function selectLastDiagnostics(diagnostics: List<Diagnostic>) {
  const max = diagnostics
    .map(d => d.step)
    .reduce((red, value) => value > red ? value : red, 0);
  return selectDiagnosticsOfStep(diagnostics, max);
}
