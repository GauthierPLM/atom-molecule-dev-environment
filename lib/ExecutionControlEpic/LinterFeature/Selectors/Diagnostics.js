"use babel";
// @flow

import type {
  Diagnostic,
  DiagnosticSeverity,
} from "../../DiagnosticsFeature/Types/types";
import type { DiagnosticsReducer } from "../../DiagnosticsFeature/Reducers/Diagnostics";
import { List, Map } from "immutable";

function isValidLocation(diagnostic: Diagnostic): boolean {
  return (
    diagnostic.path !== undefined &&
    diagnostic.path.length > 0 &&
    diagnostic.range !== undefined &&
    diagnostic.range.start !== undefined &&
    diagnostic.range.end !== undefined
  );
}

export default function DiagnosticsToLint(
  diagnostics: DiagnosticsReducer,
): List<Diagnostic> {
  return diagnostics
    .reduce(
      (acc, map) =>
        map.reduce(
          (list, severities) =>
            list
              .concat(severities.get(1) || List())
              .concat(severities.get(2) || List()),
          acc,
        ),
      List(),
    )
    .filter(isValidLocation);
}
