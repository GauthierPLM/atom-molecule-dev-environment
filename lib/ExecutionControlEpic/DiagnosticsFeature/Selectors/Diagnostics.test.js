"use babel";
// @flow

import { selectDiagnosticsOfTask } from "./Diagnostics";
import type { Diagnostic } from "../Types/types.js.flow";
import { List } from "immutable";

describe("Diagnostics Selectors", () => {
  describe("selectDiagnosticsOfTask", () => {
    it("should get diagnostics of task", () => {
      let diagnostics: List<Diagnostic> = List([
        {
          type: "success",
          message: "",
          step: 0,
          task: "2",
        },
        {
          type: "error",
          message: "",
          step: 0,
          task: "1",
        },
        {
          type: "error",
          message: "",
          step: 0,
          task: "2",
        },
        {
          type: "warning",
          message: "",
          step: 0,
          task: "2",
        },
        {
          type: "info",
          message: "",
          step: 0,
          task: "3",
        },
      ]);
      let taskId = "2";
      let subject = selectDiagnosticsOfTask(diagnostics, taskId);

      expect(subject).toMatchSnapshot();
    });

    it("should return an empty array if no diagnostics are found for task", () => {
      let diagnostics: List<Diagnostic> = List([
        {
          type: "success",
          message: "",
          step: 0,
          taskId: "3",
        },
      ]);
      let taskId = "2";
      let subject = selectDiagnosticsOfTask(diagnostics, taskId);

      expect(subject).toEqual(List());
    });
  });
});
