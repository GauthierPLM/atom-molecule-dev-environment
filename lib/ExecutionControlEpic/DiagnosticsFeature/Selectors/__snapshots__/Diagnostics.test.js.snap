// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagnostics Selectors selectDiagnosticsOfTask should get diagnostics of task 1`] = `
Immutable.List [
  Object {
    "message": "",
    "task": "2",
    "type": "success",
  },
  Object {
    "message": "",
    "task": "2",
    "type": "error",
  },
  Object {
    "message": "",
    "task": "2",
    "type": "warning",
  },
]
`;
